controllers/assignments.py

```diff
@@ -510,7 +510,11 @@ def doAssignment():
             return redirect(URL('assignments', 'chooseAssignment'))
 
     questions = db((db.assignment_questions.assignment_id == assignment.id) & \
-                   (db.assignment_questions.question_id == db.questions.id)) \
+                   (db.assignment_questions.question_id == db.questions.id) & \
```

Comment:
@presnick what is the purpose of all these extra conditions?  This is causing some people problems where custom questions they have addedto their assignments are suddenly not showing up.

Reply:
The purpose is to get the chapter name and subchapter name for readings, in order to display them rather than the no-spaces version that was displaying before.

Is the problem that questions not associated with a chapter (i.e., authored through the interface instead of an rst file) are getting filtered out by these inner joins?

One solution, though not as efficient, might be to do a separate query for each of the questions that is of type reading, to get its subchapter name.

I think it would be better to do a left outer join. We'd have to use the different web2py query syntax , as described here: http://www.web2py.com/books/default/chapter/29/06/the-database-abstraction-layer#One-to-many-relation

Reply:
I'm trying to compose and test a query right now.

Reply:
Exercises created through the web interface are stored in the sub chapter called Exercises.  We made a convention that All chapters must have an Exercises sub chapter.

So, the problem is that the StudentCSP book does not follow the convention of calling the Exercises chapter Exercises.  @ericsonga contacted me today because a bunch of her custom problems suddenly disappeared.  I couldn’t re-create the problem in any of the books I use, but when I dug deeper I saw that her book doesn’t follow the convention.  

Relying on everyone to follow conventions is risky, and this demonstrates that.

There should be a better way to ensure that every chapter has a sub chapter called Exercises.  Or else we should have a second drop down in the exercise authoring interface to make the author choose a sub chapter as well as the chapter.

Reply:
Well, I've created the following alternative query. I've tested it on my local machine and I am able to get a custom question to show up on an assignment page.

HOWEVER: after doing that, I also checked that, even with my previous query, I also go the same custom question to show up on an assignment page. So, not having reproduced the error, I'm not sure if this will fix it.

`    questions = db((db.chapters.course_id == course.course_name) & \
                   (db.sub_chapters.chapter_id == db.chapters.id) & \
                   (db.assignment_questions.assignment_id == assignment.id)) \
        .select(db.questions.name,
                db.questions.htmlsrc,
                db.questions.id,
                db.questions.chapter,
                db.questions.subchapter,
                db.assignment_questions.points,
                db.assignment_questions.activities_required,
                db.assignment_questions.reading_assignment,
                db.chapters.chapter_name,
                db.sub_chapters.sub_chapter_name,
                orderby=db.assignment_questions.sorting_priority,
                join=db.assignment_questions.on(db.questions.id == db.assignment_questions.question_id),
                left=[db.chapters.on(db.questions.chapter == db.chapters.chapter_label),
                      db.sub_chapters.on(db.questions.subchapter == db.sub_chapters.sub_chapter_label)]
                )`

Reply:
I think I need to read the docs on what it means to use the join parameter  inside select the select method.

Reply:
Here's a sample of a query generated from that, in case it helps.

```
SELECT  questions.name, questions.htmlsrc, questions.id, questions.chapter, questions.subchapter, assignment_questions.points, assignment_questions.activities_required, assignment_questions.reading_assignment, chapters.chapter_name, sub_chapters.sub_chapter_name 
FROM questions 
JOIN assignment_questions ON (questions.id = assignment_questions.question_id) 
LEFT JOIN chapters ON (questions.chapter = chapters.chapter_label) 
LEFT JOIN sub_chapters ON (questions.subchapter = sub_chapters.sub_chapter_label) 
WHERE (((chapters.course_id = 'fopp') 
AND (sub_chapters.chapter_id = chapters.id)) 
AND (assignment_questions.assignment_id = 5)
AND assignment_questions.reading_assignment IS NULL) 
ORDER BY assignment_questions.sorting_priority;
```